pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go

```diff
@@ -232,151 +300,82 @@ func encryptionKeyRotationIsSuitableControlPlane(entry *planEntry) bool {
-				Path:    encryptionKeyRotationWaitForSecretsEncryptStatusScriptPath,
-			},
-		},
-		Instructions: []plan.OneTimeInstruction{
-			p.encryptionKeyRotationRestartApiserver(cp),
-			encryptionKeyRotationWaitForSecretsEncryptStatus(cp, ""),
-			encryptionKeyRotationSecretsEncryptStatusOneTimeInstruction(cp, ""),
-		},
-		PeriodicInstructions: []plan.PeriodicInstruction{
-			p.encryptionKeyRotationApiserverStatus(cp),
-		},
+	leaderStage, err := p.encryptionKeyRotationRestartService(cp, leader, "", true)
+	if err != nil {
+		return err
 	}
-	err := assignAndCheckPlan(p.store, fmt.Sprintf("encryption key rotation: [%s]", cp.Status.RotateEncryptionKeysPhase), leader, nodePlan, 0, 0)
-	if isErrWaiting(err) {
-		if strings.HasPrefix(err.Error(), "starting") {
-			logrus.Infof("rkecluster %s/%s: restarting apiserver on %s", cp.Namespace, cp.Spec.ClusterName, leader.Machine.Name)
-			rke2.Ready.Message(cp, fmt.Sprintf("encryption key rotation: restarting apiserver for %s", leader.Machine.Name))
-			cp, err = p.rkeControlPlanes.UpdateStatus(cp)
-			if err != nil {
-				return "", p.enqueueAndSkip(cp)
-			}
+
+	logrus.Debugf("[planner] rkecluster %s/%s: collecting etcd and not control plane and not leader and init nodes", cp.Namespace, cp.Name)
+	for _, entry := range collect(clusterPlan, encryptionKeyRotationIsEtcdAndNotControlPlaneAndNotLeaderAndInit(cp)) {
+		_, err = p.encryptionKeyRotationRestartService(cp, entry, leaderStage, false)
+		if err != nil {
+			return err
 		}
-		return "", p.enqueueAndSkip(cp)
-	} else if err != nil {
-		return "", err
 	}
 
-	if _, ok := cp.Annotations[encryptionKeyRotationApiserverIDAnnotation]; !ok {
-		apiserverID := strings.TrimSuffix(string(leader.Plan.Output[encryptionKeyRotationRestartApiserverCommand]), "\n")
-		if len(apiserverID) != 13 { // not a valid container id
-			return "", p.enqueueAndSkip(cp)
-		}
-		cp.Annotations[encryptionKeyRotationApiserverIDAnnotation] = apiserverID
-		cp, err = p.rkeControlPlanes.Update(cp)
+	logrus.Debugf("[planner] rkecluster %s/%s: collecting control plane and not leader and init nodes", cp.Namespace, cp.Name)
+	for _, entry := range collect(clusterPlan, encryptionKeyRotationIsControlPlaneAndNotLeaderAndInit(cp)) {
+		stage, err := p.encryptionKeyRotationRestartService(cp, entry, leaderStage, true)
 		if err != nil {
-			return "", err
+			return err
 		}
-		return "", p.enqueueAndSkip(cp)
-	}
 
-	apiserverStatus := strings.TrimSuffix(string(leader.Plan.PeriodicOutput[encryptionKeyRotationApiserverStatusCommand].Stdout), "\n")
-	if apiserverStatus == "" || cp.Annotations[encryptionKeyRotationApiserverIDAnnotation] == apiserverStatus || len(apiserverStatus) != 13 { // not a valid container id
-		return "", p.enqueueAndSkip(cp)
+		if stage != leaderStage {
+			// secrets-encrypt command was run on another node. this is considered a failure, but might be a bit too sensitive. to be tested.
```

Comment:
Is this comment still relevant?

Reply:
Yeah. Given we're short circuiting it I'm going to leave it in